In [71]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error
%matplotlib inline

In [2]:
df = pd.read_csv('Dataset/train.csv')

In [3]:
df_test = pd.read_csv('Dataset/test.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
User_ID                       550068 non-null int64
Product_ID                    550068 non-null object
Gender                        550068 non-null object
Age                           550068 non-null object
Occupation                    550068 non-null int64
City_Category                 550068 non-null object
Stay_In_Current_City_Years    550068 non-null object
Marital_Status                550068 non-null int64
Product_Category_1            550068 non-null int64
Product_Category_2            376430 non-null float64
Product_Category_3            166821 non-null float64
Purchase                      550068 non-null int64
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [280]:
df_test.head()

User_ID Product_ID Gender    Age  Occupation City_Category  \
0  1000004  P00128942      M  46-50           7             B   
1  1000009  P00113442      M  26-35          17             C   
2  1000010  P00288442      F  36-45           1             B   
3  1000010  P00145342      F  36-45           1             B   
4  1000011  P00053842      F  26-35           1             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               1                   1   
1                          0               0                   3   
2                         4+               1                   5   
3                         4+               1                   4   
4                          1               0                   4   

   Product_Category_2  Product_Category_3  check  
0                11.0                 NaN      0  
1                 5.0                 NaN      0  
2                14.0                 NaN      0  
3                 9.0                 NaN      0  
4                 5.0                12.0      0

In [8]:
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [9]:
train_df = df.iloc[:,:-1]
train_df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  
0                 NaN                 NaN  
1                 6.0                14.0  
2                 NaN                 NaN  
3                14.0                 NaN  
4                 NaN                 NaN

In [10]:
train_df['check'] = 1
df_test['check'] = 0

In [11]:
train_df.shape, df_test.shape

((550068, 12), (233599, 12))

In [12]:
#concate the train set and test set to create 
df_concat = pd.concat([train_df, df_test], ignore_index=True)

In [13]:
df_concat

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
5       1000003  P00193542      M  26-35          15             A   
6       1000004  P00184942      M  46-50           7             B   
7       1000004  P00346142      M  46-50           7             B   
8       1000004   P0097242      M  46-50           7             B   
9       1000005  P00274942      M  26-35          20             A   
10      1000005  P00251242      M  26-35          20             A   
11      1000005  P00014542      M  26-35          20             A   
12      1000005  P00031342      M  26-35          20             A   
13      1000005  P00145042      M  26-35          20             A   
14      1000006  P00231342      F  51-55           9             A   
15      1000006  P00190242      F  51-55           9             A   
16      1000006   P0096642      F  51-55           9             A   
17      1000006  P00058442      F  51-55           9             A   
18      1000007  P00036842      M  36-45           1             B   
19      1000008  P00249542      M  26-35          12             C   
20      1000008  P00220442      M  26-35          12             C   
21      1000008  P00156442      M  26-35          12             C   
22      1000008  P00213742      M  26-35          12             C   
23      1000008  P00214442      M  26-35          12             C   
24      1000008  P00303442      M  26-35          12             C   
25      1000009  P00135742      M  26-35          17             C   
26      1000009  P00039942      M  26-35          17             C   
27      1000009  P00161442      M  26-35          17             C   
28      1000009  P00078742      M  26-35          17             C   
29      1000010  P00085942      F  36-45           1             B   
...         ...        ...    ...    ...         ...           ...   
783637  1005996  P00184242      F  26-35           0             B   
783638  1005996  P00045342      F  26-35           0             B   
783639  1006001  P00124642      F  26-35           7             A   
783640  1006002  P00121142      M  51-55           0             C   
783641  1006002  P00233542      M  51-55           0             C   
783642  1006002  P00226342      M  51-55           0             C   
783643  1006002  P00284842      M  51-55           0             C   
783644  1006003  P00313342      F  46-50          17             C   
783645  1006010  P00223542      M  36-45           0             C   
783646  1006013  P00020342      F  26-35          20             C   
783647  1006016  P00014842      M  46-50           1             B   
783648  1006016  P00109542      M  46-50           1             B   
783649  1006018  P00071442      M  36-45           1             C   
783650  1006021  P00112542      M  26-35          12             B   
783651  1006021  P00003442      M  26-35          12             B   
783652  1006023  P00262242      M  26-35           0             C   
783653  1006025  P00161142      F  26-35           1             B   
783654  1006026  P00110742      M  36-45           6             C   
783655  1006028  P00162442      M  18-25           4             A   
783656  1006030  P00129842      M  26-35          17             C   
783657  1006030  P00241442      M  26-35          17             C   
783658  1006032  P00252442      M  46-50           7             A   
783659  1006033  P00127642      M  51-55          13             B   
783660  1006035  P00135742      F  26-35           1             C   
783661  1006036  P00255442      F  26-35          15             B   
783662  1006036  P

In [14]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
User_ID                       783667 non-null int64
Product_ID                    783667 non-null object
Gender                        783667 non-null object
Age                           783667 non-null object
Occupation                    783667 non-null int64
City_Category                 783667 non-null object
Stay_In_Current_City_Years    783667 non-null object
Marital_Status                783667 non-null int64
Product_Category_1            783667 non-null int64
Product_Category_2            537685 non-null float64
Product_Category_3            237858 non-null float64
check                         783667 non-null int64
dtypes: float64(2), int64(5), object(5)
memory usage: 71.7+ MB


In [15]:
df_concat.fillna(-999, inplace=True)

In [16]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
User_ID                       783667 non-null int64
Product_ID                    783667 non-null object
Gender                        783667 non-null object
Age                           783667 non-null object
Occupation                    783667 non-null int64
City_Category                 783667 non-null object
Stay_In_Current_City_Years    783667 non-null object
Marital_Status                783667 non-null int64
Product_Category_1            783667 non-null int64
Product_Category_2            783667 non-null float64
Product_Category_3            783667 non-null float64
check                         783667 non-null int64
dtypes: float64(2), int64(5), object(5)
memory usage: 71.7+ MB


In [17]:
df_concat['Product_Category_3'].unique()

array([-999.,   14.,   17.,    5.,    4.,   16.,   15.,    8.,    9.,
         13.,    6.,   12.,    3.,   18.,   11.,   10.])

In [18]:
df_concat['User_ID'].describe()

count    7.836670e+05
mean     1.003029e+06
std      1.727267e+03
min      1.000001e+06
25%      1.001519e+06
50%      1.003075e+06
75%      1.004478e+06
max      1.006040e+06
Name: User_ID, dtype: float64

In [19]:
print(df_concat['Gender'].unique())

['F' 'M']


In [20]:
print(sorted(df_concat['Occupation'].unique()))
df_concat.groupby('Occupation')['Occupation'].count()

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


Occupation
0      98850
1      67687
2      37996
3      25126
4     103086
5      17373
6      29102
7      84127
8       2189
9       8929
10     18304
11     16593
12     44448
13     11053
14     38782
15     17356
16     36122
17     57418
18      9367
19     11919
20     47840
Name: Occupation, dtype: int64

In [21]:
print(df_concat['Age'].unique())
df_concat['Age'].describe()

['0-17' '55+' '26-35' '46-50' '51-55' '36-45' '18-25']


count     783667
unique         7
top        26-35
freq      313015
Name: Age, dtype: object

In [22]:
print(df_concat['City_Category'].unique())
print(df_concat['City_Category'].describe())

['A' 'C' 'B']
count     783667
unique         3
top            B
freq      329739
Name: City_Category, dtype: object


In [23]:
print(df_concat['Stay_In_Current_City_Years'].unique())
print(df_concat['Stay_In_Current_City_Years'].describe())
print(df_concat.groupby('Stay_In_Current_City_Years')['Stay_In_Current_City_Years'].count())

['2' '4+' '3' '1' '0']
count     783667
unique         5
top            1
freq      276425
Name: Stay_In_Current_City_Years, dtype: object
Stay_In_Current_City_Years
0     105716
1     276425
2     145427
3     135428
4+    120671
Name: Stay_In_Current_City_Years, dtype: int64


In [24]:
print(df_concat['Marital_Status'].unique())
print(df_concat['Marital_Status'].describe())
print(df_concat.groupby('Marital_Status')['Marital_Status'].count())

[0 1]
count    783667.000000
mean          0.409777
std           0.491793
min           0.000000
25%           0.000000
50%           0.000000
75%           1.000000
max           1.000000
Name: Marital_Status, dtype: float64
Marital_Status
0    462538
1    321129
Name: Marital_Status, dtype: int64


In [25]:
print(df_concat['Product_Category_1'].unique())
print(df_concat['Product_Category_1'].describe())
print(df_concat.groupby('Product_Category_1')['Product_Category_1'].count())

[ 3  1 12  8  5  4  2  6 14 11 13 15  7 16 18 10 17  9 20 19]
count    783667.000000
mean          5.366196
std           3.878160
min           1.000000
25%           1.000000
50%           5.000000
75%           8.000000
max          20.000000
Name: Product_Category_1, dtype: float64
Product_Category_1
1     200699
2      34056
3      28791
4      16756
5     215950
6      29326
7       5345
8     162294
9        604
10      7373
11     34440
12      5610
13      7930
14      2186
15      8984
16     13933
17       801
18      4436
19      1603
20      2550
Name: Product_Category_1, dtype: int64


In [26]:
print(sorted(df_concat['Product_Category_2'].unique()))
print(df_concat['Product_Category_2'].describe())
print(df_concat.groupby('Product_Category_2')['Product_Category_2'].count())
df_concat['Product_Category_2'].isnull().sum()

[-999.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    783667.000000
mean       -306.817532
std         468.194361
min        -999.000000
25%        -999.000000
50%           5.000000
75%          14.000000
max          18.000000
Name: Product_Category_2, dtype: float64
Product_Category_2
-999.0    245982
 2.0       70498
 3.0        4123
 4.0       36705
 5.0       37165
 6.0       23575
 7.0         854
 8.0       91317
 9.0        8177
 10.0       4420
 11.0      20230
 12.0       7801
 13.0      15054
 14.0      78834
 15.0      54114
 16.0      61687
 17.0      19104
 18.0       4027
Name: Product_Category_2, dtype: int64


0

In [27]:
print(sorted(df_concat['Product_Category_3'].unique()))
print(df_concat['Product_Category_3'].describe())
print(df_concat.groupby('Product_Category_3')['Product_Category_3'].count())
df_concat['Product_Category_3'].isnull().sum()

[-999.0, 3.0, 4.0, 5.0, 6.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    783667.000000
mean       -691.939130
std         465.148262
min        -999.000000
25%        -999.000000
50%        -999.000000
75%           8.000000
max          18.000000
Name: Product_Category_3, dtype: float64
Product_Category_3
-999.0    545809
 3.0         878
 4.0        2691
 5.0       23799
 6.0        6888
 8.0       17861
 9.0       16532
 10.0       2501
 11.0       2585
 12.0      13115
 13.0       7849
 14.0      26283
 15.0      39968
 16.0      46469
 17.0      23818
 18.0       6621
Name: Product_Category_3, dtype: int64


0

## Initial Analysis of features and target
* Gender has two categorical values F and M. So we will have to convert Gender from Object data type to categorical
* Occupation has 21 unique categories from 0 to 21. So it should be converted to categorical.
* Age is divided into 7 unique intervals. So convert it into categorical.
* Three categories of citys are there. So convert to categorical.
* Stay_In_Current_City_Years has 5 unique values. So convert to categorical.
* Marital Status has two values 0 and 1. So convert to categorical.
* Product_Category_1 has 20 unique values. Convert to categorical.
* Product_Category_2 had 173638 values missing specified as NaN. These converted to -999 value.
* Product_Category_3 has 383247 values missing specified as NaN. These converted to -999 value.
* Purchase is the target variable. None are missing.

Convert Product_ID and Customer_ID to categorical as well.

In [28]:
df_concat_copy = df_concat.copy()

New product category 'final_category' has a total of 237 category types which could be converted to categorical values.

In [29]:
user_id_encoder = LabelEncoder()
df_concat['User_ID'] = user_id_encoder.fit_transform(df_concat['User_ID'])

In [30]:
product_id_encoder = LabelEncoder()
df_concat['Product_ID'] = product_id_encoder.fit_transform(df_concat['Product_ID'])
age_encoder = LabelEncoder()
df_concat['Age'] = age_encoder.fit_transform(df_concat['Age'])
occupation_encoder = LabelEncoder()
df_concat['Occupation'] = occupation_encoder.fit_transform(df_concat['Occupation'])
city_category_encoder = LabelEncoder()
df_concat['City_Category'] = city_category_encoder.fit_transform(df_concat['City_Category'])
stay_encoder = LabelEncoder()
df_concat['Stay_In_Current_City_Years'] = stay_encoder.fit_transform(df_concat['Stay_In_Current_City_Years'])

In [31]:
gender_encoder = LabelEncoder()
df_concat['Gender'] = gender_encoder.fit_transform(df_concat['Gender'])

In [32]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
User_ID                       783667 non-null int64
Product_ID                    783667 non-null int64
Gender                        783667 non-null int64
Age                           783667 non-null int64
Occupation                    783667 non-null int64
City_Category                 783667 non-null int64
Stay_In_Current_City_Years    783667 non-null int64
Marital_Status                783667 non-null int64
Product_Category_1            783667 non-null int64
Product_Category_2            783667 non-null float64
Product_Category_3            783667 non-null float64
check                         783667 non-null int64
dtypes: float64(2), int64(10)
memory usage: 71.7 MB


In [33]:
df_concat.head()

User_ID  Product_ID  Gender  Age  Occupation  City_Category  \
0        0         684       0    0          10              0   
1        0        2406       0    0          10              0   
2        0         868       0    0          10              0   
3        0         844       0    0          10              0   
4        1        2769       1    6          16              2   

   Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                           2               0                   3   
1                           2               0                   1   
2                           2               0                  12   
3                           2               0                  12   
4                           4               0                   8   

   Product_Category_2  Product_Category_3  check  
0              -999.0              -999.0      1  
1                 6.0                14.0      1  
2              -999.0              -999.0      1  
3                14.0              -999.0      1  
4              -999.0              -999.0      1

In [34]:
X = df_concat.iloc[:550068,:]
y = df.iloc[:550068,-1:]

In [35]:
X = X.drop('check', axis='columns')

In [36]:
X.shape, y.shape

((550068, 11), (550068, 1))

In [37]:
X.tail()

User_ID  Product_ID  Gender  Age  Occupation  City_Category  \
550063     5883        3612       1    5          13              1   
550064     5885        3613       0    2           1              2   
550065     5886        3613       0    2          15              1   
550066     5888        3613       0    6           1              2   
550067     5889        3611       0    4           0              1   

        Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550063                           1               1                  20   
550064                           3               0                  20   
550065                           4               1                  20   
550066                           2               0                  20   
550067                           4               1                  20   

        Product_Category_2  Product_Category_3  
550063              -999.0              -999.0  
550064              -999.0              -999.0  
550065              -999.0              -999.0  
550066              -999.0              -999.0  
550067              -999.0              -999.0

In [38]:
y.tail()

Purchase
550063       368
550064       371
550065       137
550066       365
550067       490

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [40]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((385047, 11), (165021, 11), (385047, 1), (165021, 1))

In [41]:
y_test.head()

Purchase
84432      19142
72724      15513
197032      7802
353704     15455
91198       4492

In [42]:
from sklearn.tree import DecisionTreeRegressor

In [203]:
tree_reg = DecisionTreeRegressor(min_samples_split=8, min_samples_leaf = 80)

In [219]:
tree_reg.fit(X_train, y_train)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=80,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best')

In [220]:
print(tree_reg.feature_importances_)

[4.19834149e-03 7.85976828e-02 9.51609602e-04 2.17042828e-03
 1.84677138e-03 3.03043218e-03 8.56554758e-04 4.64201591e-04
 8.81308440e-01 1.42749881e-02 1.23005499e-02]


In [221]:
predictions = tree_reg.predict(X_train)

In [222]:
test_predictions = tree_reg.predict(X_test)

In [223]:
np.sqrt(mean_squared_error(y_train, predictions))

2638.805466210935

In [224]:
np.sqrt(mean_squared_error(y_test, test_predictions))

2726.052887386003

After doing some experimentations with decision tree the best validation score turn out to be 2726 when model developed
min_samples_split=8 and min_samples_leaf=80.

Now experimenting with Linear Model.

In [210]:
from sklearn.linear_model import LinearRegression

In [212]:
simple_linear = LinearRegression()
simple_linear.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [213]:
train_predicts = simple_linear.predict(X_train)
test_predicts = simple_linear.predict(X_test)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

4600.104780342809
4602.733316133697


As can be seen above even the training error is very high : 4600 without any regularization. Thus linear model clearly
underfits for this purpose.

Now lets try Random Forest.

In [215]:
from sklearn.ensemble import RandomForestRegressor

In [231]:
forest_regressor = RandomForestRegressor(min_samples_split=8, min_samples_leaf = 80, n_estimators=30)

In [232]:
forest_regressor.fit(X_train, y_train)

C:\Users\User\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=80, min_samples_split=8,
           min_weight_fraction_leaf=0.0, n_estimators=30, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [258]:
train_predicts = forest_regressor.predict(X_train)
test_predicts = forest_regressor.predict(X_test)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

2645.2995331331795
2710.2176275993893


Now running on test set provided

In [262]:
test_data = df_concat.loc[550068:,:]

In [268]:
test_data = test_data.drop('check', axis='columns')

In [270]:
predictions = forest_regressor.predict(test_data)

In [292]:
preds_df = pd.DataFrame(predictions, columns=['Purchase'])

In [296]:
preds_df.head()

Purchase
0  15793.048116
1  11560.566597
2   6195.380220
3   2532.317349
4   2622.908941

In [295]:
predictions

array([15793.04811583, 11560.56659668,  6195.38022045, ...,
       11566.33250675, 18583.69247846,  2720.99585476])

In [282]:
output = df_test.loc[:,['User_ID', 'Product_ID']]

In [300]:
output = pd.concat([output,preds_df], axis=1)

In [302]:
output.tail()

User_ID Product_ID      Purchase
233594  1006036  P00118942   7920.449656
233595  1006036  P00254642   6437.014774
233596  1006036  P00031842  11566.332507
233597  1006037  P00124742  18583.692478
233598  1006039  P00316642   2720.995855

In [304]:
output.to_csv('Submission1.csv', index=False)

As can be seen above Random Forest is performing a little bit better as compared to Decision Tree alone.

The score for test case for this random forest is 2712.6847733051.

Let us now convert some categorical to one hot vectors and see if this transformation improves the predictions or not.

In [237]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((385047, 11), (165021, 11), (385047, 1), (165021, 1))

In [238]:
X_train.head()

User_ID  Product_ID  Gender  Age  Occupation  City_Category  \
500304     4920        2020       1    2          16              0   
16861      2553        2552       1    2           1              1   
113230     5343         607       1    1           7              0   
434978      977        1200       1    2           0              0   
385516     5201         571       0    3           3              2   

        Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
500304                           0               0                   8   
16861                            2               1                  16   
113230                           1               0                   8   
434978                           1               0                   1   
385516                           3               1                   3   

        Product_Category_2  Product_Category_3  
500304                14.0              -999.0  
16861               -999.0              -999.0  
113230                16.0              -999.0  
434978                16.0              -999.0  
385516                 4.0                 5.0

In [251]:
# Convert Gender, Age, Occupation, City_Category, Stay_In_Current_Years, Marital_Status to one hot vectors.
X_train_new = pd.get_dummies(X_train, columns=['Gender', 'Age', 'Occupation','City_Category', 'Stay_In_Current_City_Years','Marital_Status'],
              drop_first=True)
X_test_new = pd.get_dummies(X_test, columns=['Gender', 'Age', 'Occupation','City_Category', 'Stay_In_Current_City_Years','Marital_Status'],
                            drop_first=True)

In [252]:
X_train_new.shape, X_test_new.shape

((385047, 39), (165021, 39))

In [253]:
tree_reg_new = DecisionTreeRegressor(min_samples_split=8, min_samples_leaf = 80)

In [254]:
tree_reg_new.fit(X_train_new, y_train)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=80,
           min_samples_split=8, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best')

In [256]:
train_predicts = tree_reg_new.predict(X_train_new)
test_predicts = tree_reg_new.predict(X_test_new)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

2642.4730930080814
2727.21621834965


No improvement observed after converting some features to one hot when using decision tree as model

Now lets try xgboost on non one hot encoded features.

In [305]:
from sklearn.ensemble import GradientBoostingRegressor

In [329]:
gradient_booster = GradientBoostingRegressor(n_estimators=1200, min_samples_leaf=80, min_samples_split=8)

In [330]:
gradient_booster

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=80,
             min_samples_split=8, min_weight_fraction_leaf=0.0,
             n_estimators=1200, presort='auto', random_state=None,
             subsample=1.0, verbose=0, warm_start=False)

In [331]:
gradient_booster.fit(X_train, y_train)

C:\Users\User\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:578: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=80,
             min_samples_split=8, min_weight_fraction_leaf=0.0,
             n_estimators=1200, presort='auto', random_state=None,
             subsample=1.0, verbose=0, warm_start=False)

In [332]:
train_predicts = gradient_booster.predict(X_train)
test_predicts = gradient_booster.predict(X_test)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

2673.146977940406
2703.673275221378
